File: AdSurgeDemo/AdSurgeDemo/ViewController.m
//
//  ViewController.m
//  AdSurgeDemo
//
//  Created by andrew taylor on 2025/3/17.
//

#import "ViewController.h"
#import "SettingManager.h"

#define kTableViewHeaderHeight 54.0
@interface ViewController () <UITableViewDelegate, UITableViewDataSource>
@property (nonatomic, strong) NSArray *demoArray;
@property (nonatomic, strong) UITableView *tableView;
@end

@implementation ViewController

- (void)loadView
{
    [super loadView];
    NSArray *section1 = @[
        @[@"RewardAD", @"RewardedViewController"],
        @[@"interstitialAD", @"InterstitialViewController"],
        @[@"BannerAD", @"BannerViewController"],
        @[@"NativeAD", @"NativeViewController"],
        @[@"AppOpenAD", @"AppOpenViewController"]];

    self.demoArray = @[
        @{@"Supported ad form": section1}
    ];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.tableView];
    self.tableView.frame = self.view.bounds;
    self.view.backgroundColor = [UIColor whiteColor];

}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self setNavigationBar];
}

- (void)setNavigationBar {
    UILabel *label = [[UILabel alloc] init];
    label.text = @"AdSurge demo - ObjC";
    label.textColor = [UIColor whiteColor];
    label.textAlignment = NSTextAlignmentCenter;
    label.font = [UIFont boldSystemFontOfSize:18];
    [label sizeToFit];
    self.navigationItem.titleView = label;

    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button setImage:[self getBtnIcon] forState:UIControlStateNormal];
    [button addTarget:self action:@selector(pressMuteBtn:) forControlEvents:UIControlEventTouchUpInside];
    button.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [button.widthAnchor constraintEqualToConstant:25],
        [button.heightAnchor constraintEqualToConstant:25]
    ]];
    UIBarButtonItem *muteBtn = [[UIBarButtonItem alloc] initWithCustomView:button];
    self.navigationItem.rightBarButtonItem = muteBtn;
}

- (void)pressMuteBtn:(UIButton*)btn {
    [SettingManager sharedManager].isMuted = ![SettingManager sharedManager].isMuted;
    UIImage *image = [self getBtnIcon];
    [btn setImage:image forState:UIControlStateNormal];
}
- (UIImage *)getBtnIcon {
    return [SettingManager sharedManager].isMuted ? [UIImage imageNamed:@"mute_icon"] : [UIImage imageNamed:@"unmute_icon"];
}
#pragma mark - UITableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.demoArray count];
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return kTableViewHeaderHeight;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section {
    return 0.0;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    UIView *header = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, kTableViewHeaderHeight)];
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(18.0, 0, header.bounds.size.width - 18.0, kTableViewHeaderHeight)];
    label.text = [[self.demoArray[section] allKeys] firstObject];
    label.textAlignment = NSTextAlignmentLeft;
    label.font = [UIFont boldSystemFontOfSize:18.0];
    [header addSubview:label];
    return header;
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    return nil;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [[[(NSDictionary *)self.demoArray[section] allValues] firstObject] count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *SimpleTableIdentifier = @"SimpleTableIdentifier";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:SimpleTableIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:SimpleTableIdentifier];
    }

    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    NSArray *cellConfig = (NSArray *)[[(NSDictionary *)self.demoArray[indexPath.section] allValues] firstObject];
    cell.textLabel.text = cellConfig[indexPath.row][0];
    UIImageView *img = [[UIImageView alloc] initWithFrame:CGRectMake(383, 15, 10, 14)];
    img.image = [UIImage imageNamed:@"showMoreBtn"];
    img.translatesAutoresizingMaskIntoConstraints = NO;
    [cell addSubview:img];
    [NSLayoutConstraint activateConstraints:@[
        [img.widthAnchor constraintEqualToConstant:10],
        [img.heightAnchor constraintEqualToConstant:14],
        [img.centerYAnchor constraintEqualToAnchor:cell.centerYAnchor],
        [img.trailingAnchor constraintEqualToAnchor:cell.trailingAnchor constant:-20]
    ]];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSArray *cellConfig = (NSArray *)[[(NSDictionary *)self.demoArray[indexPath.section] allValues] firstObject];
    id item = cellConfig[indexPath.row][1];
    if ([item isKindOfClass:[NSString class]]) {
        UIViewController *vc = [[NSClassFromString(item) alloc] init];
        [self.navigationController pushViewController:vc animated:YES];
    }
}

#pragma mark - property getter
- (UITableView *)tableView
{
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
        _tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        _tableView.delegate = self;
        _tableView.dataSource = self;
        _tableView.accessibilityIdentifier = @"tableView_id";
    }
    return _tableView;
}


@end

File: AdSurgeDemo-Swift/AdSurgeDemo-Swift/ViewController.swift
//
//  ViewController.swift
//  AdSurgeDemo-Swift
//
//  Created by andrew taylor on 2025/3/18.
//

import UIKit
import AdSurgeSDK

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    private let kTableViewHeaderHeight: CGFloat = 54.0

    private var demoArray: [[String: [[String]]]] = []
    private lazy var tableView: UITableView = {
        let tableView = UITableView(frame: self.view.bounds, style: .grouped)
        tableView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        tableView.delegate = self
        tableView.dataSource = self
        tableView.accessibilityIdentifier = "tableView_id"
        return tableView
    }()

    override func loadView() {
        super.loadView()
        let section1 = [
            ["RewardAD", "RewardedViewController"],
            ["interstitialAD", "InterstitialViewController"],
            ["BannerAD", "BannerViewController"],
            ["NativeAD", "NativeViewController"],
            ["AppOpenAD", "AppOpenViewController"],
        ]
        demoArray = [
            ["Supported ad form": section1]
        ]
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addSubview(tableView)
        tableView.frame = self.view.bounds
        self.view.backgroundColor = .white
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        setNavigationBar()
    }

    private func setNavigationBar() {
        let label = UILabel()
        label.text = "AdSurge demo - Swift"
        label.textColor = .white
        label.textAlignment = .center
        label.font = UIFont.boldSystemFont(ofSize: 18)
        label.sizeToFit()
        self.navigationItem.titleView = label

        let button = UIButton(type: .custom)
        button.setImage(getBtnIcon(), for: .normal)
        button.addTarget(self, action: #selector(pressMuteBtn(_:)), for: .touchUpInside)
        button.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            button.widthAnchor.constraint(equalToConstant: 25),
            button.heightAnchor.constraint(equalToConstant: 25)
        ])
        let muteBtn = UIBarButtonItem(customView: button)
        self.navigationItem.rightBarButtonItem = muteBtn
    }

    @objc private func pressMuteBtn(_ btn: UIButton) {
        SettingManager.shared.isMuted.toggle()
        let image = getBtnIcon()
        btn.setImage(image, for: .normal)
    }

    private func getBtnIcon() -> UIImage? {
        return SettingManager.shared.isMuted ? UIImage(named: "mute_icon") : UIImage(named: "unmute_icon")
    }

    // MARK: - UITableViewDataSource

    func numberOfSections(in tableView: UITableView) -> Int {
        return demoArray.count
    }

    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return kTableViewHeaderHeight
    }

    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 0.0
    }

    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let header = UIView(frame: CGRect(x: 0, y: 0, width: self.view.bounds.width, height: kTableViewHeaderHeight))
        let label = UILabel(frame: CGRect(x: 18.0, y: 0, width: header.bounds.width - 18.0, height: kTableViewHeaderHeight))
        label.text = Array(demoArray[section].keys)[0]
        label.textAlignment = .left
        label.font = UIFont.boldSystemFont(ofSize: 18.0)
        header.addSubview(label)
        return header
    }

    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return nil
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let values = Array(demoArray[section].values)[0]
        return values.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let SimpleTableIdentifier = "SimpleTableIdentifier"
        let cell = tableView.dequeueReusableCell(withIdentifier: SimpleTableIdentifier) ??
            UITableViewCell(style: .default, reuseIdentifier: SimpleTableIdentifier)
        cell.selectionStyle = .none
        let cellConfig = Array(demoArray[indexPath.section].values)[0]
        cell.textLabel?.text = cellConfig[indexPath.row][0]

        let img = UIImageView(frame: CGRect(x: 383, y: 15, width: 10, height: 14))
        img.image = UIImage(named: "showMoreBtn")
        img.translatesAutoresizingMaskIntoConstraints = false
        cell.addSubview(img)
        NSLayoutConstraint.activate([
            img.widthAnchor.constraint(equalToConstant: 10),
            img.heightAnchor.constraint(equalToConstant: 14),
            img.centerYAnchor.constraint(equalTo: cell.centerYAnchor),
            img.trailingAnchor.constraint(equalTo: cell.trailingAnchor, constant: -20)
        ])
        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let cellConfig = Array(demoArray[indexPath.section].values)[0]
        let className = cellConfig[indexPath.row][1]
        
        var moduleName = Bundle.main.infoDictionary?["CFBundleName"] as? String ?? ""
        moduleName = moduleName.replacingOccurrences(of: "-", with: "_")
        let fullClassName = "\(moduleName).\(className)"
        
        if let vcClass = NSClassFromString(fullClassName) as? UIViewController.Type {
            let vc = vcClass.init()
            self.navigationController?.pushViewController(vc, animated: true)
        }
    }
}

